core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java

```diff
@@ -254,22 +281,25 @@ public int getNumberOfAllocatedPages() {
 
   /**
    * Free this sorter's in-memory data structures, including its data pages and pointer array.
-   *
+   * @param bytesToReserve number of bytes to hold onto when releasing memory for this task.
    * @return the number of bytes freed.
    */
-  private long freeMemory() {
+  private long freeMemory(long bytesToReserve) {
     updatePeakMemoryUsed();
     long memoryFreed = 0;
+    long remainingBytesToReserve = bytesToReserve;
     for (MemoryBlock block : allocatedPages) {
```

Comment:
> Do we really have to release all of the data pages? 

What do you mean? That's exactly what this patch avoids. If we know we need a page immediately after the spill, we should release `n - 1` pages, where `n` is the number of pages we currently hold.